src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1132,26 +1140,17 @@ private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancell
 
         return newCompilationState;
 
-        void CheckDocumentStates<TDocumentState>(
+        static void CheckDocumentStates<TDocumentState>(
+            FilePathToDocumentIdsMap.Builder filePathToDocumentIdsMapBuilder,
             TextDocumentStates<TDocumentState> oldStates,
             TextDocumentStates<TDocumentState> newStates) where TDocumentState : TextDocumentState
         {
             if (oldStates.Equals(newStates))
```

Comment:
>oldStates [](http://example.com/codeflow?start=16&length=9)

This is a struct equality check right? A couple thoughts:

Should TextDocumentState override Equals? Doesn't seem like the list inside it should need to be checked.

Reply:
I think what's bothering me about this method is that there is there some non-obvious work being done (Equals, GetAddedStateIds, TryGetValue), and ImmutableDictionary is, umm, not performant.

What about something that (hopefully) doesn't complicate the logic, but is easier to see it's perf?

```
foreach (var (documentId, newDocumentState) in newStates.States)
{
    if (!oldStates.States.TryGetValue(documentId, out var oldDocumentState))
    {
        // Keep track of files that are definitely added.  Make sure the added doc is in the file path map.
        filePathToDocumentIdsMapBuilder.Add(newDocumentState.FilePath, documentId);
    }
    else if (oldDocumentState != newDocumentState &&
                oldDocumentState.FilePath != newDocumentState.FilePath)
    {
        // We have to check states that are in both sets as it is possible for a
        // document to change its file path without its id changing.
        filePathToDocumentIdsMapBuilder.Remove(oldDocumentState.FilePath, oldDocumentState.Id);
        filePathToDocumentIdsMapBuilder.Add(newDocumentState.FilePath, newDocumentState.Id);
    }
}
```

Reply:
> This is a struct equality check right?

No.  TextDocumentStates overrides .Equals.

Reply:
> What about something that (hopefully) doesn't complicate the logic, but is easier to see it's perf?

WFM!

Reply:
Your logic is also much better than mine for perf.  For lots of untouched projects, the newStates will be empty, meaning this loop will terminate very quickly.

Reply:
I'm still not seeing the Equals implementation for TextDocumentStates. Can you point me to it?

Reply:
```
        public bool Equals(TextDocumentStates<TState> other)
            => _map == other._map && _ids == other.Ids;
```